propertyestimator/client.py

```diff
@@ -170,40 +183,48 @@ class PropertyEstimator(object):
     >>>
     >>> # Load initial parameters
     >>> from openforcefield.typing.engines.smirnoff import ForceField
-    >>> parameters = ForceField('smirnoff99Frosst.offxml')
+    >>> parameters = ForceField(['smirnoff99Frosst.offxml'])
     >>>
-    >>> # Create a property estimator
-    >>> from propertyestimator.client import PropertyEstimator
-    >>> property_estimator = PropertyEstimator()
+    >>> ticket_id = property_estimator.estimate(data_set, parameters)
```

Comment:
I think we had wanted to always return a `PropertyEstimatorResult` instead of `ticket_id`s, since the `PropertyEstimatorResult` would be able to have code to manage either asyncrhonous or synchronous results.

Reply:
So this is something I've been conflicted about when I started looking at implementing this - In principle it seems like a really elegant solution, but in practice I'm not sure. What if for example you submit some jobs to the server, then your laptop dies. Or you submit some jobs and the server you're submitting to is super busy and can't return your results for a few day / a week etc... That PropertyEstimatorResult object or Future or whatever the return was is now lost, and you would have difficulty knowing how to ask the server for those results now.. With a simple string identifier, one can submit the jobs, wait however long it takes to get the results back, and then easily query the server about them without having to have kept the `Future` or result object alive.

Reply:
@dgasmith do you have any thoughts about this? I'm sure you must have already had to make this design decision?

Reply:
Tough call, in general even returning an ID isn't enough unless you snapshot that ID to disk somewhere. Can your clients recover state if they die?

Reply:
They can't at the moment.

Reply:
Sorry this is probably a stupid question, but what do you mean by recover state?

Reply:
ForceBalance is running along and it requests a new job and then promptly dies. When ForceBalance recovers does it know anything about the jobs that it submitted before it died? If yes, then keeping the id's around is a good id. If no, it is a bit of a moot point.

Reply:
I'm not 100% sure what ForceBalance is doing under the hood at the moment, but I'd argue that it should be saving the state of what its waiting for (so the id's). If I was to submit a large benchmark for example, I would definitely be wanting to save the state so that I could come back in a week and recover it.

Reply:
Could this be resolved by simply making `PropertyEstimatorResult` serializable?

Reply:
At the moment, ForceBalance does not know anything about the jobs that it submitted before it died.  The remote jobs that run individual simulations at different thermodynamic phase points are "workers" that connect to FB that acts as the "server".  

FB has a checkpointing system at the level of individual targets and optimization cycles.  That is to say, it can restart an optimization from saved values of the parameters, and if the latest optimization cycle has not finished evaluating the objective function, it will read the calculation results from targets where the calculations have been completed.  

If the server allows different processes to query running jobs and retrieve results, I believe it should have nice benefits but I'm not sure exactly what they would be.

Reply:
A clarification, what is the role of `PropertyEstimatorResult.id`? Is it the same as `ticket_id` in this case?

Reply:
So in both cases the id referred to is the id which the server assigns to the request. The two should be the same.

@andrrizzi, @jchodera, @j-wags I've just committed a slight refactor which tries to implement the design pattern of returning an object which itself can be asked to either synchronously or asynchronously return the results (kind of similar to returning a Future). Would you be able to take a look and let me know if this is closer to what you had in mind? The `estimate` method has been renamed to `request_estimate` and no longer takes a `synchronous` argument, as this is now delegated to the returned `Request` object. The `request_estimate` simply sends the request to the server object.

Reply:
I'm not 100% sure if I like it more than just returning an id, but I'd be happy to go with this pattern if the consensus is that it's better!

Reply:
I think this design pattern makes spinning up a server instance if one isn't running more challenging. Because the `request_estimate` method no longer has control over when the data is being returned, it won't know when to kill the server. While we could do as @andrrizzi  suggest and kill it if it's been inactive for a while, this would lead to issues if the user has waited too long to request the results of the estimate!